server/src/main/java/com/thoughtworks/go/server/service/AgentService.java

```diff
@@ -94,179 +108,186 @@ public AgentService(AgentConfigService agentConfigService, SystemEnvironment sys
-                                          List<String> resourcesToAdd, List<String> resourcesToRemove,
-                                          List<String> environmentsToAdd, List<String> environmentsToRemove,
-                                          TriState enable) {
-        agentConfigService.bulkUpdateAgentAttributes(agentInstances, username, result, uuids, environmentConfigService,
-                resourcesToAdd, resourcesToRemove, environmentsToAdd, environmentsToRemove, enable);
-    }
-
-    private boolean populateAgentInstancesForUUIDs(OperationResult operationResult, List<String> uuids, List<AgentInstance> agents) {
-        for (String uuid : uuids) {
-            AgentInstance agentInstance = findAgentAndRefreshStatus(uuid);
-            if (isUnknownAgent(agentInstance, operationResult)) {
-                return false;
+    public void bulkUpdateAgentAttributes(List<String> uuids, List<String> resourcesToAdd, List<String> resourcesToRemove,
+                                          EnvironmentsConfig envsToAdd, List<String> envsToRemove, TriState state,
+                                          LocalizedOperationResult result) {
+        AgentsUpdateValidator validator = new AgentsUpdateValidator(agentInstances, uuids, state, resourcesToAdd, resourcesToRemove, result);
+        try {
+            if (isAnyOperationPerformedOnBulkAgents(resourcesToAdd, resourcesToRemove, envsToAdd, envsToRemove, state, result)) {
+                validator.validate();
+
+                List<Agent> agents = agentDao.getAgentsByUUIDs(uuids);
+                if (isTriStateSet(state)) {
+                    agents.addAll(agentInstances.filterPendingAgents(uuids));
+                }
+
+                agents.forEach(agent -> setResourcesEnvsAndState(agent, resourcesToAdd, resourcesToRemove, envsToAdd, envsToRemove, state));
+                updateIdsAndGenerateCookiesForPendingAgents(agents, state);
+                agentDao.bulkUpdateAttributes(agents, state);
+                result.setMessage("Updated agent(s) with uuid(s): [" + join(uuids, ", ") + "].");
+            }
+        } catch (Exception e) {
+            LOGGER.error("There was an error bulk updating agents", e);
+            if (!result.hasMessage()) {
+                result.internalServerError("Server error occured. Check log for details.");
             }
-            agents.add(agentInstance);
         }
-        return true;
     }
 
-    public void deleteAgents(Username username, HttpOperationResult operationResult, List<String> uuids) {
-        if (!hasOperatePermission(username, operationResult)) {
-            return;
-        }
-        List<AgentInstance> agents = new ArrayList<>();
-        if (!populateAgentInstancesForUUIDs(operationResult, uuids, agents)) {
+    public void updateAgentsAssociationOfEnvironment(EnvironmentConfig envConfig, List<String> uuids, LocalizedOperationResult result) {
```

Comment:
Seems to be used only in tests.

Reply:
Used in Environments SPA. See `environments_controller.rb:63`.